src/c2goto/library/pthread_lib.c

```diff
@@ -486,20 +501,37 @@ __ESBMC_HIDE:;
 int pthread_key_create(pthread_key_t *key, void (*destructor)(void *))
 {
 __ESBMC_HIDE:;
-  // TODO
-  return 0;
+  __ESBMC_atomic_begin();
+  __ESBMC_assert(
+    key != NULL,
+    "In pthread_key_create, key parameter should be different than NULL.");
+  __ESBMC_thread_keys[__ESBMC_next_thread_key] = NULL;
+  __ESBMC_thread_key_destructors[__ESBMC_next_thread_key] = destructor;
+  *key = __ESBMC_next_thread_key++;
+  __ESBMC_atomic_end();
+  return nondet_bool() ? (nondet_bool() ? EAGAIN : ENOMEM) : 0;
```

Comment:
The model here is wrong: despite succeeding in creating the key, we still might return errors? 

The [documentation](https://pubs.opengroup.org/onlinepubs/007904875/functions/pthread_key_create.html) says:
```
The pthread_key_create() function shall fail if:

[EAGAIN]
The system lacked the necessary resources to create another thread-specific data key, or the system-imposed limit on the total number of keys per process {PTHREAD_KEYS_MAX} has been exceeded.
[ENOMEM]
Insufficient memory exists to create the key.
```

There are public implementations of this function that we can use as base for our model: [here](https://android.googlesource.com/platform/external/pthreads/+/216eb8153b2455d1e15f8fbf84d4413ddcc40adc/pthread_key_create.c)

We should either:
1. Correctly model all the errors that may happen.
2. Assume it always succeeds, as we do for other models in the pthread lib.

In the case of 1, the code should be something like:
```
  atomic_begin();
  int result = 0;
  pthread_key_t newkey = (pthread_key_t) calloc (1, sizeof (*newkey));

  if (newkey == NULL)
    {
      result = ENOMEM;
    }
  else if (/* Check if we can insert the key in the array */)
    {
      result = EAGAIN;

      free (newkey);
      newkey = NULL;
    }
  else if (destructor != NULL)
    {
      newkey->keyLock = 0;
      newkey->destructor = destructor;
    }

  *key = newkey;

  atomic_end();
  return result;
```